Applications/Spire/Source/KeyBindings/CancelKeyBindingsTableModel.cpp

```diff
@@ -1,61 +1,125 @@
+        return Action::ALL;
+    }
+  }
+
+  constexpr auto get_index(Action action) {
+    switch(action) {
+      case get_action(0):
+        return 0;
+      case get_action(1):
+        return 1;
+      case get_action(2):
+        return 2;
+      case get_action(3):
+        return 3;
+      case get_action(4):
+        return 4;
+      case get_action(5):
+        return 5;
+      case get_action(6):
+        return 6;
+      case get_action(7):
+        return 7;
+      case get_action(8):
+        return 8;
+      case get_action(9):
+        return 9;
+      case get_action(10):
+        return 10;
+      case get_action(11):
+        return 11;
+      case get_action(12):
+        return 12;
+      default:
+        return -1;
+    }
   }
 
-  auto get_index(KeyBindings::CancelAction action) {
-    static auto indexes = std::unordered_map<KeyBindings::CancelAction, int>({
-      {KeyBindings::CancelAction::ALL, 0},
-      {KeyBindings::CancelAction::ALL_ASKS, 1},
-      {KeyBindings::CancelAction::ALL_BIDS, 2},
-      {KeyBindings::CancelAction::CLOSEST_ASK, 3},
-      {KeyBindings::CancelAction::CLOSEST_BID, 4},
-      {KeyBindings::CancelAction::FURTHEST_BID, 5},
-      {KeyBindings::CancelAction::FURTHEST_ASK, 6},
-      {KeyBindings::CancelAction::MOST_RECENT, 7},
-      {KeyBindings::CancelAction::MOST_RECENT_ASK, 8},
-      {KeyBindings::CancelAction::MOST_RECENT_BID, 9},
-      {KeyBindings::CancelAction::OLDEST, 10},
-      {KeyBindings::CancelAction::OLDEST_ASK, 11},
-      {KeyBindings::CancelAction::OLDEST_BID, 12}});
-    return indexes[action];
+  constexpr auto get_action_text(int row) {
+    switch(row) {
+      case 0:
+        return std::string_view("All");
+      case 1:
+        return std::string_view("All Asks");
+      case 2:
+        return std::string_view("All Bids");
+      case 3:
+        return std::string_view("Closest Ask");
+      case 4:
+        return std::string_view("Closest Bid");
+      case 5:
+        return std::string_view("Furthest Ask");
+      case 6:
+        return std::string_view("Furthest Bid");
+      case 7:
+        return std::string_view("Most Recent");
+      case 8:
+        return std::string_view("Most Recent Ask");
+      case 9:
+        return std::string_view("Most Recent Bid");
+      case 10:
+        return std::string_view("Oldest");
+      case 11:
+        return std::string_view("Oldest Ask");
+      case 12:
+        return std::string_view("Oldest Bid");
+      default:
+        return std::string_view("Invalid cancel action");
+    }
   }
 }
 
 CancelKeyBindingsTableModel::CancelKeyBindingsTableModel(
-    const std::vector<KeyBindings::CancelActionBinding>& bindings,
-    QObject* parent)
-    : QAbstractTableModel(parent) {
+    const std::vector<Binding>& bindings, QObject* parent)
+    : QAbstractTableModel(parent),
+      m_key_bindings(ROW_COUNT) {
   for(auto i = 0; i < ROW_COUNT; ++i) {
     m_key_bindings.push_back({{}, {}, get_action(i)});
```

Comment:
I think it's a bug now. Prior to the loop, `m_key_bindings` has `ROW_COUNT` default-constructed elements. Inside the loop, you are pushing additional elements to the end of the vector.
You can replace this line with
`m_key_bindings[i] = {...};`

Reply:
I mixed that up and thought that constructor was just for reserving space. So empty elements are preferred to using reserve? Or should I just remove the explicit construction and call reserve in the constructor body here?

Reply:
CancelActionBinding's default constructor doesn't do anything so it does not really matter. I would probably go with `reserve` just to impose fewer requirements on `CancelActionBinding` (no assignment, copy constructor only).